pkg/operator/sync.go

```diff
@@ -1240,6 +1240,19 @@ func (optr *Operator) reconcileMachineOSBuilder(mob *appsv1.Deployment) error {
 		return fmt.Errorf("could not get layered MachineConfigPools: %w", err)
 	}
 
+	if len(layeredMCPs) > 0 {
+		nodes, err := optr.GetAllManagedNodes(layeredMCPs)
+		if err != nil {
+			return fmt.Errorf("could not get nodes in layered MachineConfigPools: %w", err)
+		}
+		// Error out if any of the nodes in the pool where OCL is enabled is a non-coreos node
+		for _, node := range nodes {
+			if !strings.Contains(node.Status.NodeInfo.OSImage, "Red Hat Enterprise Linux CoreOS") && !strings.Contains(node.Status.NodeInfo.OSImage, "Fedora CoreOS") && !strings.Contains(node.Status.NodeInfo.OSImage, "CentOS Stream CoreOS") {
+				return fmt.Errorf("on cluster layering is only supported on CoreOS based nodes; detected node OS %q", node.Status.NodeInfo.OSImage)
```

Comment:
issue (non-blocking): If a cluster has multiple non-CoreOS nodes, a cluster admin will encounter this error message once for each different node, which can be frustrating.

suggestion (non-blocking): Instead, we could identify all non-CoreOS nodes and return an aggregate error which contains individual errors for each non-CoreOS node:

```golang
// ...

errNodes := []error{}
for _, node := range nodes {
    if !isCoreOSNode(node) {
        err := fmt.Errorf("non-CoreOS OS (%s) detected on node %s in MachineConfigPool(s) %v", node.Status.NodeInfo.OSImage, node.Name, getMachineConfigPoolsForNode(node))
        errNodes = append(errNodes, node)
    }
}

if len(errNodes) != 0 {
    // aggerrs is here: https://github.com/kubernetes/apimachinery/blob/master/pkg/util/errors/errors.go#L46
    return fmt.Errorf("on-cluster layering is only supported on CoreOS-based nodes: %w", aggerrs.NewAggregate(errNodes))
}

// ...
```

These are non-blocking because I don't know how common this scenario actually is. But given that we supported bring-your-own-RHEL (non-RHCOS)-nodes, it's definitely non-zero. However, I will leave this to your discretion as well :smile:

Reply:
Thank you for the pointers! Fixed to aggregate the errors and add more info to the errors